File: 백준/Gold/3190. 뱀/뱀.py
import sys
from collections import deque
input = sys.stdin.readline

n = int(input())
k = int(input())
apple = []

for _ in range(k):
    apple.append(list(map(int, input().split())))

l = int(input())

move = deque()
for _ in range(l):
    move.append(list(input().split()))

dxy = ((0, 1), (1, 0), (0, -1), (-1, 0))
dir = 0
way = dxy[dir]

second = 0

snake = deque()
snake.append([1, 1])

while True:
    # 시간 증가
    second += 1

    # 덱의 앞에서 push하고 뒤에서 Pop하는 방식으로 진행
    # 머리 위치를 바꾸지 말고 진행 방향에 있는 블럭을 추가해줌

    # 머리 이동
    movingX = snake[0][0]
    movingY = snake[0][1]
    movingX += way[0]
    movingY += way[1]
    snake.appendleft((movingX, movingY))
    
    # print(f'moving = {moving[0]} | moving = {moving[1]}')
    # 맵 영역 밖
    if movingX < 1 or movingX > n or movingY < 1 or movingY > n:
        print(second)
        quit()

    # 머리, 몸통 만남
    for i in range(1, len(snake)):
        if snake[0][0] == snake[i][0] and snake[0][1] == snake[i][1]:
            print(second)
            quit()
    tail = snake.pop()  # 꼬리 이동
    # 사과 먹음
    for i in range(len(apple)):
        if apple and movingX == apple[i][0] and movingY == apple[i][1]:
            del apple[i]
            snake.append(tail)
            break

    # 시간에 따른 방향 확인
    if move and second == int(move[0][0]):
        if move[0][1] == 'D':
            dir = (dir + 1) % 4
        else:
            dir = (dir + 3) % 4
        way = dxy[dir]
        move.popleft()
    #print(f'second = {second} | {snake}')
    # 위치 확인용
    # print(f'second = {second} / moving = {moving} / snake = {snake}')

File: 백준/Gold/3190. 뱀/뱀.cc
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

int board[101][101]; // 빈칸: 0, 사과: 1, 뱀: 2
int n, k;
int sec = 1; // 진행 시간
int dir = 0; // 우0, 하1, 좌2, 상3

int dx[] = {0, 1, 0, -1};
int dy[] = {1, 0, -1, 0};

deque<pair<int, int>> snake;
queue<pair<int, char>> movement;

void input(){
    cin >> n >> k;
    for(int i = 0; i < k; i++){
        int x, y;
        cin >> x >> y;
        board[x-1][y-1] = 1;
    }

    int l;
    cin >> l;
    for(int i = 0; i < l; i++){
        int a;
        char c;
        cin >> a >> c;
        movement.push({a, c});
    }
    snake.push_back({0, 0});
    board[0][0] = 2;
}

void print(){
    for(int i = 0; i < n; i++){
        for(int j = 0; j < n; j++){
            cout << board[i][j] << ' ';
        }
        cout << endl;
    }
}

void start(){
    while(10){
        int nx = snake.front().first + dx[dir];
        int ny = snake.front().second + dy[dir];
        if(nx < 0 || ny < 0 || nx >= n || ny >= n || board[nx][ny] == 2) break; // 벽 or 뱀
        if(board[nx][ny] == 0){ // 빈칸
            board[snake.back().first][snake.back().second] = 0;
            snake.pop_back();
            board[nx][ny] = 2;
            snake.push_front({nx, ny});
        }
        else if(board[nx][ny] == 1){ // 사과
            snake.push_front({nx, ny});
            board[nx][ny] = 2;
        }
        if(movement.front().first == sec){ // 진행 방향 변경
            if(movement.front().second == 'D') dir++; // 오른쪽 회전
            else dir--; // 왼쪽 회전
            dir = (dir+4) % 4;
            movement.pop();
        }
        sec++;
    }
}

int main(){
    input();
    start();
    cout << sec << endl;
}